UnitTestFrameworkPkg/Library/Posix/MemoryAllocationLibPosix/MemoryAllocationLibPosix.c

```diff
@@ -265,21 +273,24 @@ FreeAlignedPages (
   )
 {
   PAGE_HEAD  *PageHeadPtr;
+  VOID       *AllocatedBuffer;
+  UINTN      Length;
 
-  //
-  // NOTE: Partial free is not supported. Just keep it.
-  //
-  PageHeadPtr = (VOID *)((UINTN)Buffer - sizeof (PAGE_HEAD));
-  if (PageHeadPtr->Signature != PAGE_HEAD_PRIVATE_SIGNATURE) {
-    return;
-  }
+  ASSERT (Buffer != NULL);

```

Comment:
We should try to move away from relying on asserts for validating potential null pointers.

Reply:
Why? This validates a function precondition, not an operation result.

Reply:
What is the advantage of the function continuing to process the pointer value if asserts are disabled and the pointer is null?

Reply:
The pointer is never NULL, that is the precondition. In nicer languages, you can enforce this with the type system, but when you’re stuck with C, asserts is all you got. If you started checking for NULL in every function that takes a pointer - well, I wouldn’t want to see the diff (or the increase in code size). In the real world, the choice whether to check for NULL or not, in my experience, is absolutely arbitrary and inconsistent.

My rationale why it doesn’t make sense for free functions is here: https://github.com/tianocore/edk2/pull/6408/files/00791b2efb0641452dde1530623790e0d24474d0#r1920858384

Reply:
We can agree to disagree on this point. I understand the concern about redundant checks and optimization. But, in a language like C, where the input can be anything regardless of assumptions made about the calling code, I think each function that is part of an external interface is responsible for ensuring the safety of system execution by validating its parameters.

Reply:
But you can’t. So someone doesn’t pass NULL but 0xAFAFAFAF… now what?

Reply:
After verifying the pointer is not null, verifying that `PAGE_HEAD_PRIVATE_SIGNATURE` is valid would be reasonable (as the code did previously).

Reply:
NULL is not *the* invalid pointer, it’s the *intentionally invalid* pointer. There are many invalid pointers and many of them will fault because they’re not mapped or because of alignment violations. Alignment is something that’s basically not checked anywhere in EDK II at all, sometimes it’s even violated by code that’s supposed to be correct. Point being, defensive NULL checks are sort of a fallacy for validity, while really that’s one case of arbitrarily many you cannot possible check for.

Reply:
This is tracking further from the original comment. I did not say NULL is the only type of invalid pointer. I did not say that alignment is also not an issue in code.

The original comment was about code that solely uses an assert to test an error condition with the note that asserts can be disabled (and often are in production systems). In that case, the system would be more reliable if that assert were to not further execute with _an_ invalid pointer (that is acknowledged by the developer as invalid in the scenario per the assert).

Reply:
No, you didn’t say any of that. You said functions should validate pointers to ensure safety and I gave examples for why they cannot. They *have* to trust the caller with validity, no way around it. NULL is one completely arbitrary way it can be broken - and it’s a case that’s unlikely to be hit even accidentally due to caller constraints.

Reply:
I understand that many things can go wrong. I agree with you on that. The original comment is not a statement about the spectrum of potential pitfalls in memory management or the probability of this function being given a null pointer (we've seen cases like this in pre-existing binaries with null pointer detection enabled).

It is that the function has an assert right now to identify a condition that is invalid, and that check will be absent in some production platforms where those systems would be more robust if it were not absent. Stepping back though, that's less of a concern with this instance in UnitTestFrameworkPkg.

In any case, this was a good conversation, and I'll check the `OcGuardLib` you mentioned. Mike, feel free to do what you like with this thread :).

Reply:
I would prefer to move this conversation to its own GitHub Issue.  This topic has been discussed many times before and there is a lot of good input from the community from those previous conversations.